daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -544,25 +544,27 @@ private final class `Value Equal instance`[Cid: Equal] extends Equal[Value[Cid]]
   import Value._
   import ScalazEqual._
 
-  override final def equalIsNatural: Boolean = Equal[Cid].equalIsNatural
+  override final def equalIsNatural: Boolean = false
 
   implicit final def Self: this.type = this
 
   override final def equal(a: Value[Cid], b: Value[Cid]) =
     (a, b).match2 {
       case a @ (_: ValueInt64 | _: ValueNumeric | _: ValueText | _: ValueTimestamp | _: ValueParty |
           _: ValueBool | _: ValueDate | ValueUnit) => { case b => a == b }
-      case r: ValueRecord[Cid] => { case ValueRecord(tycon2, fields2) =>
+      case r: ValueRecord[Cid] => { case ValueRecord(_, fields2) =>
```

Comment:
Is this because we assume that we enforce that only equal types are compared? If so, I think this deserves a comment on the method.

Reply:
> we are using this equality for value of different type ?

This was originally defined for validation. So I think the answer to your question is _yes_, since data that is invalid in that way may be supplied, by a malicious or mistaken actor.

As such, I would make this a tagged equality.

```
private[data] final class `Value SameType Equal instance`[Cid] extends Equal[Value[Cid]] {
...
}

implicit def `Value SameType Equal instance`[Cid: Equal]: Equal[Value[Cid] @@ SameType] =
  Tag.subst(new `Value SameType Equal instance`)
```